datafusion/functions-window/src/row_number.rs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines physical expression for `row_number` that can evaluated at runtime during query execution
+
+use std::any::Any;
+use std::fmt::Debug;
+use std::ops::Range;
+
+use datafusion_common::arrow::array::ArrayRef;
+use datafusion_common::arrow::array::UInt64Array;
+use datafusion_common::arrow::compute::SortOptions;
+use datafusion_common::arrow::datatypes::DataType;
+use datafusion_common::{Result, ScalarValue};
+use datafusion_expr::expr::WindowFunction;
+use datafusion_expr::{Expr, PartitionEvaluator, Signature, Volatility, WindowUDFImpl};
+
+/// Create a [`WindowFunction`](Expr::WindowFunction) expression for
+/// `row_number` user-defined window function.
+pub fn row_number(args: Vec<Expr>) -> Expr {
```

Comment:
We will have a macro `create_func` similar to aggregate later on, right?

Reply:
Yes definitely.

Reply:
BTW I don't think `row_number()` takes arguments  -- I removed the argument in https://github.com/apache/datafusion/pull/12043


This is the previous row_number implementation here https://github.com/apache/datafusion/pull/12030/files#diff-790fda7313189b10bd29f89102416b93f5708ab88fdd5c89733758e1ed2aafadL22

```rust
/// Create an expression to represent the `row_number` window function
pub fn row_number() -> Expr {
    Expr::WindowFunction(WindowFunction::new(
        BuiltInWindowFunction::RowNumber,
        vec![],
    ))
}
```

Reply:
@alamb Thank you for taking care of this in #12043.